base/timing.jl

```diff
@@ -195,21 +199,54 @@ julia> @time begin
 """
 macro time(ex)
     quote
-        while false; end # compiler heuristic: compile this block (alter this if the heuristic changes)
-        local stats = gc_num()
-        local compile_elapsedtime = cumulative_compile_time_ns()
-        local elapsedtime = time_ns()
-        local val = $(esc(ex))
-        elapsedtime = time_ns() - elapsedtime
-        compile_elapsedtime = cumulative_compile_time_ns() - compile_elapsedtime
-        local diff = GC_Diff(gc_num(), stats)
-        time_print(elapsedtime, diff.allocd, diff.total_time,
-                   gc_alloc_count(diff), compile_elapsedtime)
+        # First, measure how long it takes to compile the @time block without a body
+        local empty_body_compiletime = cumulative_compile_time_ns()
+        $(_time_expr(nothing))
+        empty_body_compiletime = cumulative_compile_time_ns() - empty_body_compiletime
+
+        # Then, run the real code, and measure the compilation time there, and subtract out
+        # the time for just the overhead.
+        local outer_compiletime = cumulative_compile_time_ns()
+        elapsedtime, inner_compiletime, gcdiff, val = $(_time_expr(ex))
+        outer_compiletime = cumulative_compile_time_ns() - outer_compiletime
+
+        # Subtract out the inner (dynamic dispatch) compilation time to get toplevel + overhead.
+        outer_compiletime = outer_compiletime - inner_compiletime
+        # Threshold to display compilation time (because computers have hella variance)
+        overhead_ratio = outer_compiletime / empty_body_compiletime
+        if overhead_ratio > 2.0  # three standard deviations of noise in our experiments
+            # Subtract out the overhead, so we reflect only the true compiletime for user code.
+            toplevel_compiletime = outer_compiletime - empty_body_compiletime
+        else
+            toplevel_compiletime = 0
+        end
+
+        time_print(elapsedtime, gcdiff.allocd, gcdiff.total_time,
+                   gc_alloc_count(gcdiff), inner_compiletime, toplevel_compiletime)
+
         println()
         val
     end
 end
 
+function _time_expr(ex)
+    quote
+        Base.inferencebarrier(()->begin
+            while false; end # compiler heuristic: compile this block (alter this if the heuristic changes)
+            local stats = gc_num()
+            local compiletime = cumulative_compile_time_ns()
+            local elapsedtime = time_ns()
+            local val = $(esc(ex))
+            elapsedtime = time_ns() - elapsedtime
+            compiletime = cumulative_compile_time_ns() - compiletime
+            local gcdiff = GC_Diff(gc_num(), stats)
+            elapsedtime, compiletime, gcdiff, val
+        end)()
```

Comment:
@yuyichao: I think this addresses the concern you raised in https://github.com/JuliaLang/julia/pull/37938#discussion_r501815151, about `Does this change using @time in local scope?`.

This means (I think) that every call to `@time` would add a non-inlined function call, but that seems worth it so that we can report accurate compilation times?

What do you think? Thanks!